File: Sliding Window/1. Best Time to Buy and Sell Stock/solution.exs
defmodule Solution do
  @spec max_profit(prices :: [integer]) :: integer
  def max_profit(prices) do
    recursive(prices, 100_000, 0)
  end

  def recursive([], _, result), do: result
  def recursive([last], min_left, result), do: max(result, last - min_left)

  def recursive([left, right | tail], min_left, result) do
    new_min_left = min(left, min_left)
    recursive([right | tail], new_min_left, max(result, right - new_min_left))
  end
end

File: Sliding Window/1. Best Time to Buy and Sell Stock/solution.scala
object Solution {
    def maxProfit(prices: Array[Int]): Int = {
        var min_left = 100000
        var result = 0
        for(i <- 0 to prices.length - 1) {
            min_left = min(min_left, prices(i))
            result = max(result, prices(i) - min_left)
        }
        result
    }

    def max(x: Int, y: Int): Int = {
        if (x > y) x
        else y
    }

    def min(x: Int, y: Int): Int = {
        if (x < y) x
        else y
    }
}
